OPTM-55 Recursive crdt

guisedlythe opened the pull request:
Preliminary implemenation of recursive Task merging using the `recursive-crdt` branch in `lww_crdt`.

There is quite some additional code in the TaskListService.

guisedlythe committed:
OPTM-74 Added a (failing) test to demonstrate the ClockDriftException

guisedlythe committed:
OPTM-74 Implemented a fix for the ClockDriftException.

- If the exception occurs, backdate all incoming HLC timestamps that are too far in the future to the largest possible value and try to merge again

guisedlythe committed:
OPTM-74 Added tests to check if HLC-based merge still works when there is no ClockDriftException

guisedlythe committed:
OPTM-55 Use own crdt library instead of the flutter crdt library

guisedlythe committed:
OPTM-55 Changed github CI to use recursive clone

guisedlythe committed:
Merge branch 'OPTM-74-crdt-clock-drift' into OPTM-55-crdt

guisedlythe committed:
OPTM-55 Added method to TaskListService to get task records with peer ids and timestamps

guisedlythe committed:
OPTM-55 Added basic recursive crdt for tasks.

- Based on MapCrdtNode in lww_crdt.
- Task members variables are still atomic (either used after merge or not).
- Removed storing of Task.id in the crdt value because it is already used as the key and therefore would only be a duplication.
- Updated/Added test to check if a simple example works.
- Updated minimum dart version to 2.13 (and version of dart_code_metrics to 4.0 because there is a conflict otherwise). This enables typedefs for non-function types.

guisedlythe committed:
OPTM-55 Updated llw_crdt

guisedlythe committed:
OPTM-55 Use putAll where applicable

guisedlythe committed:
OPTM-55 Merge remote-tracking branch 'origin/main' into OPTM-55-recursive-crdt

!! There are still some very hacky workarounds !!

- Tasks are now contained in TasksList s
- Every property of the TaskList is stored in the CRDT and synced (i.e. isShared, title, sortBy)
- Because everything is stored in one CRDT, only one messages is needed to sync tasks and task lists
- Updated/Adjusted tests

guisedlythe committed:
OPTM-55 Fixed crdt node name change

This should never happen anyway...

guisedlythe committed:
OPTM-55 Added TaskListService.allTaskRecords. Changed some getters in task list service to private visibility.

guisedlythe committed:
OPTM-55 Use DeviceTaskList helper class in task_list_service_test.dart

- Fixes tests failing because the database is not reset after each test

guisedlythe committed:
OPTM-55 Fixed small typo in a comment

firstdayofjune commented on lib/models/task_list.dart:
```diff
@@ -14,16 +14,18 @@ part 'task_list.g.dart';
     this.id,
     required this.title,
-    this.elements,
+    this.elements = const [],
```
This line looks dangerous to me. Is the `const []` evaluated whenever a task list is created or only at compile time? Because in the second case all TaskLists will share the same elements-list.

guisedlythe replied:
It actually creates a constant list of size 0 that can not be changed (you also cannot add new elements). Still, this can be very confusing and I will change it.

firstdayofjune commented on lib/screens/task_list_screen.dart:
```diff
@@ -134,17 +135,17 @@ class _TaskListScreenState extends State<TaskListScreen> {
-            listen: false,
-          ).callbackProvider;
-          listService.upsert(widget.taskList);
+          taskListService.upsertTaskList(
```
Sorting tasks does not work anymore. The icon next to the sort option toggles, but the tasks are not displayed in the new order.

firstdayofjune commented on lib/services/task_list_service.dart:
```diff
@@ -23,141 +27,340 @@ class TaskListService with LogMixin, ChangeCallbackProvider {
-  }
+  Future<Iterable<TaskRecord>> get allTaskRecords async =>
+      _decodeTasks(await _readAndDecodeTaskListCollectionCrdt()).map(
+        (v) => TaskRecord(
```
what is v?

firstdayofjune commented on test/services/task_list_service_test.dart:
```diff
@@ -1,47 +1,79 @@
+    tasks = (await device.taskListService.allTasks).toList();
+    expect(tasks.isEmpty, true);
+
+    await device.taskListService.upsertTask('id1', Task(title: taskTitle2));
```
You tested whether you can create and remove a task in the first part of the test. Why do you test the create/retrieve again? What could break by deleting a task?

guisedlythe replied:
Deleting a task only marks the task as deleted (the task becomes a tombstone). I want to test if it can get changed back to an existing task.

firstdayofjune replied:
Understood. I would split this test into several tests then (also considering my comment above).
So write one test for each:

- test task id is set after creation
- test task can be retrieved from the database
- test task can be deleted from the database (and is marked as tombstone)
- test deleted task is recovered by creating it again

To avoid copy-pasting the task creation, either create the tasks in a setup method or in a helper.
Also, you could consider grouping the tasks (https://pub.dev/packages/test#writing-tests)

firstdayofjune commented on test/services/task_list_service_test.dart:
```diff
@@ -1,47 +1,79 @@
+    final allTaskRecords =
+        (await device.taskListService.allTaskRecords).toList();
+    expect(allTaskRecords.length, 2);
+    expect(allTaskRecords.map((v) => v.task).toSet(), {task1, task2});
```
Is the order of the set guaranteed? Could you instead assert that task1 and task2 is contained in the allTaskRecords list?

guisedlythe replied:
It's a set, it doesn't have an order.
I do this to check that the tasks are contained and that they are the only tasks.

firstdayofjune replied:
Ok, I wasn't sure how the comparison behaves.
Let's use the [containsAll](https://api.dart.dev/stable/2.13.4/dart-core/Set/containsAll.html) method here to make this more explicit:

`allTaskRecords.toSet().containsAll({task1, task2})`

(With allTaskRecords being a [list](https://api.dart.dev/stable/2.13.4/dart-core/List-class.html), `.toSet()` should work without the map.)

guisedlythe replied:
The `map` is necessary because `allTaskRecords` is of type `List<TaskRecord>`.
I think using sets for this kind of comparison is very readable and intuitive, but I agree that the previous line 58 is unnecessary and I will remove it. Using `containsAll` only checks if all elements are contained, but not if they are the only elements (in combination with line 58 it would be equivalent though).

guisedlythe replied:
See same problem as above

firstdayofjune commented on test/services/task_list_service_test.dart:
```diff
@@ -1,47 +1,79 @@
+        (await device.taskListService.allTaskRecords).toList();
+    expect(allTaskRecords.length, 2);
+    expect(allTaskRecords.map((v) => v.task).toSet(), {task1, task2});
+    expect(allTaskRecords.map((v) => v.taskListId).toList(), ['id1', 'id1']);
```
if this is possible, I would rather do something like:
```
allTaskRecords.map((record) => expect(record.taskListId, "id1"));
```

firstdayofjune commented on test/services/task_list_service_test.dart:
```diff
@@ -1,47 +1,79 @@
+    var tasks = (await device.taskListService.allTasks).toList();
+    expect(tasks.length, 1);
+    expect(tasks.first.title, taskTitle1);
+    final taskId = tasks.first.id!;
```
Shouldn't the ID be created by the Task(...) constructor already? I would assert the ID is set in a separate test, because IMO this does not belong to the store/retrieve/delete context.

firstdayofjune commented on test/services/task_list_service_test.dart:
```diff
@@ -1,47 +1,79 @@
+    final allTaskRecords =
+        (await device.taskListService.allTaskRecords).toList();
+    expect(allTaskRecords.length, 1);
+    expect(allTaskRecords.first.task, null);
```
What is the point in keeping the task as a null value in the database after removal?

guisedlythe replied:
Its a tombstone. The underlying implementation does not know that the task was never synced to other devices (it would not actually need a tombstone in this case).

The allTaskRecords should also return tombstones so we can display them as deleted in the activity log.

firstdayofjune commented on lib/services/task_list_service.dart:
```diff
@@ -23,141 +27,340 @@ class TaskListService with LogMixin, ChangeCallbackProvider {
+
+  Future<Iterable<Task>> getTasksFromList(String taskListId) async =>
+      getTaskListById(taskListId)
+          .then((taskList) => taskList != null ? taskList.elements : []);
```
Depending on the usage, I would consider returning null for non-existing taskListIDs. The current implementation hides this error from the developer and could make debugging harder (because an existing, but empty task list behaves the same like the non-existing list).

firstdayofjune commented on lib/services/task_list_service.dart:
```diff
@@ -23,141 +27,340 @@ class TaskListService with LogMixin, ChangeCallbackProvider {
+  Future<Iterable<Record<TaskList>>> get _taskListRecords async =>
+      (await _taskListRecordMap).values;
 
-    switch (taskList.sortBy) {
```
This was the sorting logic 😄 Did you implement this anywhere else? Otherwise this might have broken the now missing sort functionality. 🙃

firstdayofjune commented on lib/services/task_list_service.dart:
```diff
@@ -23,141 +27,340 @@ class TaskListService with LogMixin, ChangeCallbackProvider {
+
+    task.id ??= Uuid().v4();
+    final taskNode = taskListCrdt.getRecord(task.id!);
+    if (taskNode == null || taskNode.isDeleted) {
```
If this does not become to excessive, could you add a docstring explaining what a taskNode is / how the CRDT is structured? Currently, I don't understand why taskNode == null and its deletion is handled the same. (Obviously I don't understand this, because I don't know the CRDT functionality in detail. So if it is possible to summarize this in a docstring/comment would be great!)

firstdayofjune commented on lib/services/task_list_service.dart:
```diff
@@ -23,141 +27,340 @@ class TaskListService with LogMixin, ChangeCallbackProvider {
+    final update = (await _readAndDecodeTaskListCollectionCrdt())
+      ..get(taskListId)?.delete(taskId);
+    await _store(update);
+    await _syncService.run();
```
Does the syncService then decide whether to actually sync or not (e.g. based on the auto-sync configuration)?

firstdayofjune commented on test/services/task_list_merge_test.dart:
```diff
@@ -0,0 +1,189 @@
+    ];
+  });
+
+  test('distinct task lists', () async {
```
is this `test('multiple task lists keep their tasks distinct', ...` or test multiple task lists don't mix their tasks

guisedlythe replied:
I agree that this test is not very good. It is there to test if the task lists that are created in setup() are really distinct, but it is more a test of the memory database than a test of the task list merging.

I originally added this test because I got unexpected errors at the beginning caused the databases from the different test devices not being distinct.

Should I just delete this whole test or move it into a separate file?

Solution: Move this test into separate file. Explain why it is necessary (e.g. on Windows wrong setup can lead to the use of the same db for different devices)

guisedlythe committed:
OPTM-55 Readded task and task list sorting

- Sorting is now in the screen code and not in the service
- Accidentally removed when removing TaskListsService

firstdayofjune commented on test/services/task_list_merge_test.dart:
```diff
@@ -0,0 +1,189 @@
+    expect(taskLists.first.title, 'list');
+    expect(taskLists.first.elements.toSet(), {task1, task2});
+    expect(
+      (await devices[0].taskListService.getTasksFromList('id')).toSet(),
```
Does toSet() guarantee the order to always be the same?

guisedlythe replied:
=> Add test to check that a task is not contained more than once in the task list.

firstdayofjune commented on test/services/task_list_merge_test.dart:
```diff
@@ -0,0 +1,189 @@
+      (await devices[0].taskListService.getTasksFromList('id')).toSet(),
+      {task1, task2},
+    );
+    expect((await devices[0].taskListService.allTasks).toSet(), {task1, task2});
```
How about adding another expect to check if devices[1] also has multiple tasks now?

guisedlythe replied:
This test only merges into device[0], so device[1] should still be unchanged.

firstdayofjune commented on test/services/task_list_merge_test.dart:
```diff
@@ -0,0 +1,189 @@
+    final task2 = Task(id: task1.id, title: 'task2');
+    await devices[0].taskListService.upsertTaskList(taskList);
+    await devices[0].taskListService.upsertTask(taskList.id!, task1);
+    await Future.delayed(Duration(milliseconds: 10));
```
Is this wait actually needed or could we instead manipulate a timestamp of a Task?

guisedlythe replied:
Yeah, but this is easier and takes less lines of code. You actually shouldn't even need to wait because every operation takes time, this is just to be extra safe (in case it takes less than one ms). It also illustrates what this test is all about.

I get that using a delay in a test is not good style, but in this case I think that manipulating the timestamp is not worth it (and can introduce additional bugs). Also, it's just 10ms...

firstdayofjune replied:
If we don't need it, remove the wait completely.

If we do, let's test the logic and make it reproducible by explicitly adding a time delta to the second task.
More lines of code is neither bad nor good. If you think, the extra lines make the test more difficult to understand, create a helper method (e.g. `incrementTimestamp(task2, Duration(milliseconds: 1));`). This is explicit and we can reuse it in other tests as well.

Adding the delay introduces a code smell to our code base, which is likely to be copied by other developers when adding further tests.

guisedlythe replied:
Removing it completely would, in theory, make it possible for the test to fail if it executes too quickly (less than 1ms between the insertion of both tasks).
The best I would agree to do is to change the delay to 1ms, which is the minimum for the test to be guaranteed to succeed.

**About manually adjusting the timestamp:**
It is currently not possible to  change the timestamp of tasks using the task list service because the service abstracts the underlying crdt implementation away. It is still be possible if you manipulate the json in the storage, but that is very hacky and totally dependent on the task list service implementation. This is done in the  'crdt merge clock drift' test where I even added comments that this is not a good test. I only did it there because it the only way to test this behavior on one device if you don't turn back your wall clock.
Another option would be to add a method to the task list service that allows for this timestamp manipulation. This would, however, only be useful in tests. Do you think this is worth the time investment?

firstdayofjune commented on test/services/task_list_merge_test.dart:
```diff
@@ -0,0 +1,189 @@
+  test('crdt update less recent task in list', () async {
+    final taskList = TaskList(id: 'id', title: 'list');
+    final task1 = Task(id: 'task1id', title: 'task1');
+    final task2 = Task(id: task1.id, title: 'task2');
```
If there is no reason for `task1.id` please just use the same string for both tasks. This made me think. I don't want to think 😂

firstdayofjune commented on test/services/task_list_merge_test.dart:
```diff
@@ -0,0 +1,189 @@
+    final task2 = Task(id: task1.id, title: 'task2');
+    await devices[1].taskListService.upsertTaskList(taskList);
+    await devices[1].taskListService.upsertTask(taskList.id!, task2);
+    await Future.delayed(Duration(milliseconds: 10));
```
Same as above, can we instead set explicit timestamps? Not only does this "slow down" the tests, delays like this may also create non-reproducible test failures.

firstdayofjune commented on test/services/task_list_merge_test.dart:
```diff
@@ -0,0 +1,189 @@
+      description: 'description1',
+    );
+    final task2 = Task(
+      id: task1.id,
```
Please just reuse the ID, e.g. use `"SameID"` for both tasks.

firstdayofjune commented on test/services/task_list_merge_test.dart:
```diff
@@ -0,0 +1,189 @@
+        .taskListService
+        .mergeCrdtJson(await devices[1].taskListService.crdtToJson());
+    expect((await devices[0].taskListService.allTasks).toSet(), {
+      Task(
```
This test can easily fail, when the Task-Constructor changes (e.g. when adding a `created` timestamp). Check for the respective values instead (so compare the description and the title explicitly).

firstdayofjune reviewed (changes requested):
Sorting functionality is now broken, except for that only some minor things 👍 
(I did not understand the CRDT implementation completely, so I did not comment anything there.)

guisedlythe committed:
OPTM-55 Implemented some things from PR review

guisedlythe committed:
OPTM-55 Refactored some task list service tests according to PR review

guisedlythe committed:
OPTM-55 Added TODO comment and changed delay in tests to 1ms according to PR review

guisedlythe committed:
OPTM-55 Updated lww_crdt library

- The library update contains a bug fix that *might* fix the recursive sync bug we encountered

guisedlythe committed:
Merge branch 'main' into OPTM-55-recursive-crdt

- Made SyncService optional in most services to make sure it doesn't interfere with other tests
- Also fixed a typo in a crdt test

firstdayofjune reviewed (approved):
Looks good! Only some final remarks (minor refactoring issues) we will cover in a separate PR as discussed.

guisedlythe commented on test/services/task_list_merge_test.dart:
```diff
@@ -0,0 +1,258 @@
+    expect(await devices[1].taskListService.getTasksFromList('id2'), [task2]);
+  });
+
+  test('crdt merge tasks unordered same list', () async {
```
Specify this (and other tests) that use one-way merge explicitly in test title.
